File: backend-java/src/test/java/com/skewwhiffy/auraltester/notation/factory/NoteFactoryTest.java
package com.skewwhiffy.auraltester.notation.factory;

import com.skewwhiffy.auraltester.notation.model.key.Key;
import com.skewwhiffy.auraltester.notation.model.note.AbsoluteNote;
import lombok.val;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.junit.jupiter.MockitoExtension;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

@ExtendWith(MockitoExtension.class)
public class NoteFactoryTest {
    @InjectMocks
    private NoteFactory noteFactory;

    @Test
    void instantiatesMiddleC() {
        val expected = AbsoluteNote.getMiddleC();

        val actual = noteFactory.getAbsoluteNote("C");

        assertThat(actual).isEqualTo(expected);
    }

    @Test
    void instantiatesNoteAboveMiddleC() {
        val expected = "c''";

        val actual = noteFactory.getAbsoluteNote(expected);

        assertThat(actual.getAbc(Key.getCMajor())).isEqualTo(expected);
    }

    @Test
    void instantiatesNoteBelowMiddleC() {
        val internalNotation = "Dx#,,,";
        val expected = "^^^D,,,";

        val actual = noteFactory.getAbsoluteNote(internalNotation);

        assertThat(actual.getAbc(Key.getCMajor())).isEqualTo(expected);
    }

    @Test
    void when_noteNameInvalid_then_throws() {
        assertThatThrownBy(() -> noteFactory.getAbsoluteNote("H"))
                .isInstanceOf(IllegalArgumentException.class);
    }
}

File: backend.old/src/test/kotlin/com/skewwhiffy/auraltester/notation/factory/NoteFactoryTest.kt
package com.skewwhiffy.auraltester.notation.factory

import com.skewwhiffy.auraltester.notation.model.key.Key
import com.skewwhiffy.auraltester.notation.model.note.AbsoluteNote
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest

@SpringBootTest
class NoteFactoryTest {
    @Autowired
    private lateinit var noteFactory: NoteFactory

    @Test
    fun instantiatesMiddleC() {
        val expected = AbsoluteNote.middleC
        val actual = noteFactory.getAbsoluteNote("C")
        assertThat(actual).isEqualTo(expected)
    }

    @Test
    fun instantiatesNoteAboveMiddleC() {
        val expected = "c''"
        val actual = noteFactory.getAbsoluteNote(expected)
        assertThat(actual.getAbc(Key.cMajor)).isEqualTo(expected)
    }

    @Test
    fun instantiatesNoteBelowMiddleC() {
        val internalNotation = "Dx#,,,"
        val expected = "^^^D,,,"
        val actual = noteFactory.getAbsoluteNote(internalNotation)
        assertThat(actual.getAbc(Key.cMajor)).isEqualTo(expected)
    }

    @Test
    fun when_noteNameInvalid_then_throws() {
        assertThatThrownBy { noteFactory.getAbsoluteNote("H") }
            .isInstanceOf(IllegalArgumentException::class.java)
    }
}

